twill-core/src/main/java/org/apache/twill/internal/json/ResourceReportAdapter.java

```diff
@@ -59,4 +72,61 @@ public ResourceReport fromJson(String json) {
   public ResourceReport fromJson(Reader reader) {
     return gson.fromJson(reader, ResourceReport.class);
   }
+
+  /**
+   * A type adapter for serializing {@code Collection<TwillRunResources>} correctly
+   */
+  private static final class ResourceReportTypeAdapterFactory implements TypeAdapterFactory {
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
+      if (!Collection.class.isAssignableFrom(type.getRawType())) {
+        return null;
```

Comment:
This factory is specifically for serializing`Collection<TwillRunResources>` since Gson will not catch the inner class for it. So we only need to create the TypeAdapter when we encounter `Collection<TwillRunResources>`. Actually the implementation is simulated from `TwillRuntimeSpecificationAdapter`, which create factory to deal with similar situation, https://github.com/apache/twill/blob/master/twill-core/src/main/java/org/apache/twill/internal/json/TwillRuntimeSpecificationAdapter.java#L106